### PR TITLE
doc/dev: Add internal documentation for mclock scheduler

### PR DESCRIPTION
**Pull request by sseshasa**
The document details the refinements to mclock scheduler and presents the
results of a comparison study performed between the mclock scheduler and
the WPQ scheduler.

Signed-off-by: Sridhar Seshasayee <sseshasa@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst [hunk 1]
```diff
@@ -0,0 +1,488 @@
...
+
+
+Recovery Test Steps
+===================
+
+Before bringing up the Ceph cluster, the following mClock configuration
+parameters were set appropriately based on the obtained baseline throughput
+from the previous section:
+
+- ``osd_mclock_max_capacity_iops_hdd``
```
Comment (tchaikov) [inline doc/dev/osd_internals/mclock_wpq_cmp_study.rst:214]: nit, better off referencing these settings using 
```rst
:confval:`osd_mclock_max_capacity_iops_hdd`
```

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst [hunk 2]
```diff
@@ -0,0 +1,488 @@
...
+Ceph's use of mClock was primarily experimental and approached with an
+exploratory mindset. This is still true with other organizations and individuals
+who continue to either use the codebase or modify it according to their needs.
+
+DmClock exists in its own repository_. Before the Ceph *Pacific* release,
+mClock could be enabled by setting the ``osd_op_queue`` Ceph option to
+*mclock_scheduler*. Additional mClock parameters like *reservation*, *weight*
+and *limit* for each service type could be set using Ceph options.
+For example, ``osd_mclock_scheduler_client_[res,wgt,lim]`` is one such option.
+See `OSD Config Reference`_ section for more details. Even with all the mClock
```
Comment (tchaikov) [inline doc/dev/osd_internals/mclock_wpq_cmp_study.rst:14]: better off using
```rst
:ref:`dmclock-qos`
```

the relative link is fragile.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 1 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst:214]: @tchaikov Sure, I will change it here and in other places.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 2 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst:14]: Sure, I will push this in my next update.

**Review by sunnyku — APPROVED** [review]
LGTM

**Review by neha-ojha — APPROVED** [review]
The content looks good! 

@zdover23 After this PR is merged, could you please give a read to doc/dev/osd_internals/mclock_wpq_cmp_study.rst and suggest refinements in the form of follow-on PRs.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+================
+
+Ceph cbt_ was used to test the recovery scenarios. A new recovery test to
+generate background recoveries with client I/Os in parallel was created.
+See the next section for the detailed test steps. The test was executed 3 times
+with the default *Weighted Priority Queue (WPQ)* scheduler for comparison
+purposes. This was done to establish a credible mean value to compare
+the mClock scheduler results at a later point.
+
+Post this, the same test was executed with mClock scheduler and with different
```
Comment (jdurgin) [inline doc/dev/osd_internals/mclock_wpq_cmp_study.rst:71]: nit: 'after' is more conventional than 'post' as a preposition in american english

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst [hunk 4]
```diff
@@ -0,0 +1,488 @@
+=========================================
+ QoS Study with mClock and WPQ Schedulers
+=========================================
+
```
Comment (jdurgin) [inline doc/dev/osd_internals/mclock_wpq_cmp_study.rst:4]: could use an intro to what qos / mclock is intended to do at a high level, maybe referencing the existing docs for this background

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+overwhelm the client operations. Sections further below discuss the recovery
+rates under these conditions.
+
+With the non-default options, the same test was executed with mClock and with
+the default profile(*high_client_ops*) enabled. As per the profile allocation,
+the reservation goal of 50% (10750 IOPS) is being met with an average throughput
+of 11209 IOPS during the course of recovery operations. This is more than 4x
+times the throughput obtained with WPQ(BST).
+
+Similar throughput with the *balanced*(11017 IOPS) and *high_recovery_ops*
```
Comment (jdurgin) [inline doc/dev/osd_internals/mclock_wpq_cmp_study.rst:294]: I think you need a space before the ( - it's throwing off the formatting in the rendered version

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst [hunk 6]
```diff
@@ -0,0 +1,488 @@
...
+
+.. image:: ../../images/mclock_wpq_study/Recovery_Rate_Comparison_HDD_NoWALdB_WPQ_vs_mClock.png
+
+Key Takeaways and Conclusion
+============================
+
+- mClock is able to provide the desired QoS using profiles to allocate proper
+  *reservation*, *weight* and *limit* to the service types.
+- By using the cost per I/O and the cost per byte parameters, mClock can perform
+  the tag calculations for the different device types(SSD/HDD).
```
Comment (jdurgin) [inline doc/dev/osd_internals/mclock_wpq_cmp_study.rst:479]: tags aren't really a focus of this doc, so it's confusing to mention them when the reader may not be familiar with the mclock algorithm

```suggestion
- By using the cost per I/O and the cost per byte parameters, mClock can schedule operations appropriately for the different device types(SSD/HDD).
```

**Review by jdurgin — CHANGES_REQUESTED** [review]
looks great overall, very detailed and comprehensive description! left a few comments to address

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 4 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst:4]: Done. Please verify.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 3 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst]: Ack. I have changed it.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 4 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst:4]: Have added an "Introduction" section and referenced the appropriate documentation including the mClock paper.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 5 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst]: Fixed. Thanks for pointing this out.

File: doc/dev/osd_internals/mclock_wpq_cmp_study.rst (hunk 6 above)
Comment (sseshasa) [reply doc/dev/osd_internals/mclock_wpq_cmp_study.rst]: I have changed it according you suggestion. I have removed the mention of tags in other places as well.

**Comment by sseshasa** [comment]
jenkins test docs

**Comment by sseshasa** [comment]
enkins test docs

**Comment by sseshasa** [comment]
jenkins test docs

**Comment by sseshasa** [comment]
jenkins test docs

**Review by jdurgin — APPROVED** [review]
Looks good, thanks for adding the intro!

**Review by tchaikov — APPROVED** [review]
lgtm from the formatting perspective. =)

**Pull request merged**